server/src/main/java/org/elasticsearch/common/util/ReleasableDoubleArray.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.apache.lucene.util.RamUsageEstimator;
+import org.elasticsearch.common.bytes.ReleasableBytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+class ReleasableDoubleArray implements DoubleArray {
+    private static final long SHALLOW_SIZE = RamUsageEstimator.shallowSizeOfInstance(ReleasableDoubleArray.class);
+
+    private final ReleasableBytesReference ref;
+
+    ReleasableDoubleArray(StreamInput in) throws IOException {
+        ref = in.readReleasableBytesReference();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeBytesReference(ref);
+    }
+
+    @Override
+    public long size() {
+        return ref.length() / 8;
+    }
+
+    @Override
+    public double get(long index) {
+        if (index > Integer.MAX_VALUE / 8) {
+            // We can't serialize messages longer than 2gb anyway
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        return ref.getDoubleLE((int) index * 8);
+    }
+
+    @Override
+    public double set(long index, double value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public double increment(long index, double inc) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void fill(long fromIndex, long toIndex, double value) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void set(long index, byte[] buf, int offset, int len) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long ramBytesUsed() {
+        /*
+         * If we return the size of the buffer that we've sliced
+         * we're likely to double count things.
+         */
+        return SHALLOW_SIZE;
+    }
+
+    @Override
+    public void close() {
+        ref.decRef();
```

Comment:
Maybe this is obvious, but I'm not used to looking at this part of the code - where's the corresponding `incRef` for this?

Reply:
You'll love this. It's in the ctor - in `readReleasableBytesReference`. The way this links into the rest of the world is that `ReleasableBytesReference#streamInput` returns a specialized input that incs the ref when you call `readReleasableBytesReference`.